src/Beacon.jsx

```diff
@@ -1,111 +1,206 @@
+    const appRoot = document.querySelector(`.${TOOLTIP_OVERLAY_CLASS}`);
+    const appRootClassName = appRoot && appRoot.className;
+    this.setState({ persistent, indexedDB, appRoot, appRootClassName });
+
+    if (persistent) {
+      // Retrieve the state of the badge from the database
+      // If `persistent` is `true` then we use the hash of the component's children
+      // as a unique ID. `persistent` can be set to some other truthy value to override this default ID.
+      const hash = persistent === true ? sha1(JSON.stringify(this.props.tooltipText)) : persistent;
+      const request = indexedDB.open('react-beacon');
+      request.onupgradeneeded = (event) => {
+        const db = event.target.result;
+        db.createObjectStore('beacons');
+      };
+      request.onsuccess = () => {
+        this.setState({ database: request.result });
+        const transaction = this.state.database.transaction(['beacons']);
+        const objectStore = transaction.objectStore('beacons');
+        objectStore.get(hash).onsuccess = (event) => {
+          if (!event.target.result) {
+            this.setState({ hash });
+          }
+        };
+      };
+    }
   }
 
   componentDidMount() {
-    const { parent } = this.props;
-    this.setState({ parentEl: document.querySelector(parent) });
+    this.targetElement = this.getTargetElement();
   }
 
-  componentDidUpdate(prevProps) {
-    if (prevProps.parent !== this.props.parent) {
-      const { parent } = this.props;
-      this.setState({ parentEl: document.querySelector(parent) });
+  componentWillUpdate() {
+    // If margins are similar then center tooltip
+    if (this.targetElement) {
+      const bounds = this.targetElement.getBoundingClientRect();
+      const centerTooltip = Math.abs(window.innerWidth - bounds.right - bounds.left) < TOOLTIP_TOLERANCE;
+      if (this.state.centerTooltip !== centerTooltip) {
+        this.setState({ centerTooltip });
+      }
     }
   }
 
-  getCoordinates(position, bounds) {
-    switch (position) {
-    case 'top':
-      return {
-        left: bounds.left + ((bounds.right - bounds.left - BEACON_WIDTH) / 2),
-        top: bounds.top
-      };
-    case 'bottom':
-      return {
-        left: bounds.left + ((bounds.right - bounds.left - BEACON_WIDTH) / 2),
-        top: bounds.bottom - BEACON_HEIGHT
-      };
-    case 'right':
-      return {
-        left: bounds.right - BEACON_WIDTH,
-        top: bounds.top + ((bounds.bottom - bounds.top - BEACON_HEIGHT) / 2)
-      };
-    default:
-    case 'left':
-      return {
-        left: bounds.left,
-        top: bounds.top + ((bounds.bottom - bounds.top - BEACON_HEIGHT) / 2)
-      };
+  componentDidUpdate() {
+    this.targetElement = this.getTargetElement();
+  }
+
+  getTargetElement() {
+    if (!this.refs.tooltipTarget) {
+      return null;
+    } else if (this.refs.tooltipTarget instanceof React.Component) {
+      return ReactDOM.findDOMNode(this.refs.tooltipTarget);
```

Comment:
OK so I really hate this... I would use `div` wrapper in the render function instead.